### PR TITLE
website/docs/cloud-docs/integrations/kubernetes: Update migration guide

### PR DESCRIPTION
**Pull request by arybolovlev**
### What

Update the HCP Terraform Operator migration guide with the CRD migration table, change the order of some steps, and update them whenever necessary.

### Why

https://hashicorp.atlassian.net/browse/TFECO-6425

### Screenshots

N/A.

----------

### Merge Checklist
_If items do not apply to your changes, add (N/A) and mark them as complete._

#### Pull Request
- [ ] One or more labels describe the type of change (e.g. clarification) and associated product (e.g. HCP Terraform ).
- [ ] Description links to related pull requests or issues, if any.

#### Content
- [ ] Redirects have been added to `website/redirects.js` for moved, renamed, or deleted pages.
- [ ] API documentation and the API Changelog have been updated. 
- [ ] Links to related content where appropriate (e.g., API endpoints, permissions, etc.).
- [ ] Pages with related content are updated and link to this content when appropriate.
- [ ] Sidebar navigation files have been updated for added, deleted, reordered, or renamed pages.
- [ ] New pages have metadata (page name and description) at the top.
- [ ] New images are 2048 px wide. They have HashiCorp standard annotation color (#F92672) and format (rectangle with rounded corners), blurred sensitive details (e.g. credentials, usernames, user icons), and descriptive alt text in the markdown for accessibility.
- [ ] New code blocks have the correct syntax and line breaks to eliminate horizontal scroll bars.
- [ ] UI elements (button names, page names, etc.) are bolded.
- [ ] The Vercel website preview successfully deployed.

#### Reviews
- [ ] I or someone else reviewed the content for technical accuracy.
- [ ] I or someone else reviewed the content for typos, punctuation, spelling, and grammar.

### CODE REVIEW

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 1]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+  name: operator
 ```
 
-Change to the migration directory.
+The table below details the mapping between the `Workspace` controller in v1 of the operator and the `Module` controller in v2 of the operator.
+
+| Version 1 (Workspace CRD) | Version 2 (Module CRD) | Notes |
+| --- | --- | --- |
+| `spec.module` | N/A | In v2 of the operator a `Module` is a separate controller with its own CRD. |
+| N/A | `spec.name: operator` | In v1 of the operator, the name of the generated module is hardcoded to `operator`. In v2, the default name of the generated module is `this`, however it can be changed. |
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:165]:
```suggestion
| N/A | `spec.name: operator` | In v1 of the operator, the name of the generated module is hardcoded to `operator`. In v2, the default name of the generated module is `this`, but you can rename it. |
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 2]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `spec.module.source` | `spec.module.source` | This supports all Terraform [module sources](https://developer.hashicorp.com/terraform/language/modules/sources). |
+| `spec.module.version` | `spec.module.version` | Refer to  [module sources](https://developer.hashicorp.com/terraform/language/modules/sources) for versioning information for each module source. |
+| `spec.variables.[*]` | `spec.variables.[*].name` | Variable names should be included in the module. This is a reference to variables in the Workspace where the module will be executed. |
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:168]: We only capitalize product names:
```suggestion
| `spec.variables.[*]` | `spec.variables.[*].name` | You should include variable names in the module. This is a reference to variables in the workspace that is executing the module. |
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 3]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `spec.outputs.[*].key` | `spec.outputs.[*].name` | Output names should be included in the module. This is a reference to the output variables produced by the module. |
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:169]:
```suggestion
| `spec.outputs.[*].key` | `spec.outputs.[*].name` | You should include output names in the module. This is a reference to the output variables produced by the module. |
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 4]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `status.workspaceID` OR `metadata.namespace-metadata.name` | `spec.workspace.id` OR `spec.workspace.name` | The workspace where the module will be executed. The workspace must be within the same organization. |
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:170]: In the same organization as what? Or are you referring to listing multiple workspace names and those workspaces must be in the same org?
```suggestion
| `status.workspaceID` OR `metadata.namespace-metadata.name` | `spec.workspace.id` OR `spec.workspace.name` | The workspace where the module is executed. The workspace must be in the same organization. |
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 5]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+    - key: SECRET_KEY
+      value: "s3cr3t"
+      hcl: false
+      sensitive: false
+      environmentVariable: true
+```
+
+</CodeBlockConfig>
+
+In v2 of the operator, the workspace and module are manage by separate controllers.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:200]:
```suggestion
In v2 of the operator, separate controllers manage workspace and modules.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 6]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+## Upgrade the operator
+
+Download Workspace CRD patch A:
 
 ```shell-session
-$ cd terraform-cloud-operator/docs/migration/crds
+$ curl -sO https://raw.githubusercontent.com/hashicorp/terraform-cloud-operator/main/docs/migration/crds/workspaces_patch_a.yaml
 ```
 
 View the changes that **patch A** will apply to the workspace CRD.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:250]: We only use bold for UI elements:
```suggestion
View the changes that patch A applies to the workspace CRD.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 7]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
 
 ```shell-session
-$ kubectl diff -f workspaces_patch_a.yaml
+$ kubectl diff --filename workspaces_patch_a.yaml
 ```
 
 Patch the workspace CRD with **patch A**. This patch adds `app.terraform.io/v1alpha2` support, but excludes `.status.runStatus` because it has a different format in `app.terraform.io/v1alpha1` and causes JSON unmarshalling issues.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:256]:
```suggestion
Patch the workspace CRD with patch A. This patch adds `app.terraform.io/v1alpha2` support, but excludes `.status.runStatus` because it has a different format in `app.terraform.io/v1alpha1` and causes JSON un-marshalling issues.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 8]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
 
+~> **Warning** Once the patch is applied, Kubernetes will convert existing `app.terraform.io/v1alpha1` custom resources to `app.terraform.io/v1alpha2` according to the updated schema and v1 of the operator will no longer serve custom resources. You must update existing custom resources to satisfy the current schema. Refer to the tables above for a mapping of resource fields between v1 and v2 of the operator.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:258]: I think we can shorten this to:
```suggestion
!> **Upgrade warning**: Once you apply a patch, Kubernetes converts existing `app.terraform.io/v1alpha1` custom resources to `app.terraform.io/v1alpha2` according to the updated schema, meaning that v1 of the operator can no longer serve custom resources. Before patching, update your existing custom resources to satisfy the v2 schema requirements. [Learn more](#manifest-schema-migration).
```

I also suggested using the [suggested callout styling for upgrade warnings](https://github.com/hashicorp/engineering-docs/blob/main/writing/style-guide.md#labels)

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 9]
```diff
@@ -94,28 +280,25 @@ Next, create a Kubernetes secret to store the HCP Terraform API token following
 $ kubectl --namespace ${RELEASE_NAMESPACE} get secret terraformrc -o json | jq '.data.credentials' | tr -d '"' | base64 -d
 ```
 
-If v2 of the operator pod fails to start, ensure that each `workspace` object has the `spec.token` value set. To do this, patch each `workspace` object using following command for each Workspace resource.
+Update existing custom resources accoring to the migration table above and apply your changes.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:283]:
```suggestion
Update existing custom resources [according to the schema migration guidance](#manifest-schema-migration) and apply your changes.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 10]
```diff
@@ -94,28 +280,25 @@ Next, create a Kubernetes secret to store the HCP Terraform API token following
...
 ```shell-session
-$ kubectl logs -f <POD_NAME>
-INFO	Workspace Controller	{"workspace": {"name":"<WORKSPACE_NAME>","namespace":"<NAMESPACE>"}, "msg": "new reconciliation event"}
-##...
-INFO	Workspace Controller	{"workspace": {"name":"<WORKSPACE_NAME>","namespace":"<NAMESPACE>"}, "msg": "successfully reconciled workspace"}
+$ curl -sO https://raw.githubusercontent.com/hashicorp/terraform-cloud-operator/main/docs/migration/crds/workspaces_patch_b.yaml
 ```
 
-View the changes that **patch B** will apply to the workspace CRD:
+View the changes that **patch B** will apply to the workspace CRD.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:295]:
```suggestion
View the changes that patch B applies to the workspace CRD.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 11]
```diff
@@ -94,28 +280,25 @@ Next, create a Kubernetes secret to store the HCP Terraform API token following
...
 
 ```shell-session
-$ kubectl diff -f workspaces_patch_b.yaml
+$ kubectl diff --filename workspaces_patch_b.yaml
 ```
 
-Patch workspace CRD with **patch B**. This patch adds `.status.runStatus` support that was excluded in Patch A.
+Patch workspace CRD with **patch B**. This patch adds `.status.runStatus` support that was excluded in patch A.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:301]:
```suggestion
Patch workspace CRD with patch B. This patch adds `.status.runStatus` support, which was excluded in patch A.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 12]
```diff
@@ -132,14 +315,28 @@ Specifically, look for an error message such as the following.
...
 ```
+
 The `finalizer` exists to provide greater control over the migration process. Verify the custom resource, and when you’re ready to migrate it, use the `kubectl patch` command to update the `finalizer` value.
 
 ```shell-session
 $ kubectl patch workspace migration --type=merge --patch '{"metadata": {"finalizers": ["workspace.app.terraform.io/finalizer"]}}'
 ```
 
-Finally, review the operator logs once more and verify there are no error messages reported.
+Review the operator logs once more and verify there are no error messages reported.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:325]:
```suggestion
Review the operator logs once more and verify there are no error messages.
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 13]
```diff
@@ -132,14 +315,28 @@ Specifically, look for an error message such as the following.
...
 
 ```shell-session
 $ kubectl logs -f <POD_NAME>
-```
\ No newline at end of file
+```
+The operator will reconcile resources during the next sync period. This interval is set by the `operator.syncPeriod` configuration of the operator and defaults to five minutes.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:330]:
```suggestion
The operator reconciles resources during the next sync period. This interval is set by the `operator.syncPeriod` configuration of the operator and defaults to five minutes. 
```

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 14]
```diff
@@ -132,14 +315,28 @@ Specifically, look for an error message such as the following.
...
+```
+The operator will reconcile resources during the next sync period. This interval is set by the `operator.syncPeriod` configuration of the operator and defaults to five minutes. 
+
+If you have any migrated `Module` custom resources, apply them now.
+
+```shell-session
+$ kubectl apply --filename <MIGRATED_V2_MODULE_MANIFEST.yaml>
+```
+
+In v2 of the operator, `applyMethod` is set to `manual` by default. In this case, a new run in a managed workspace will require manual approval. Run the following command for each Workspace resource to change it to `auto` approval.
```
Comment (rkoron007) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:338]:
```suggestion
In v2 of the operator, the `applyMethod` is set to `manual` by default. In this case, a new run in a managed workspace requires manual approval. Run the following command for each `Workspace` resource to change it to `auto` approval.
```

**Review by rkoron007 — APPROVED** [review]
Left style suggestions, but overall is looking great! As always, please feel free to push back on anything that doesn’t feel good to you! ✨

**Review by BrianMMcClain — APPROVED** [review]

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 15]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
-Finally, we suggest that you set the `applyMethod` of each HCP Terraform workspace to `manual`. In version 2 of the Terraform Cloud Kubernetes Operator this is the default value, however explicitly configuring it with the value provides extra protection from unintended automatic applies during the upgrade process. Run the following command for each Workspace resource.
+## Manifest schema migration
+
+In v2 of the operator, many fields have been renamed or moved. You must update your specification with these changes when you migrate to v2 of the operator.
+
+### Workspace controller
+
+Below is the field mapping of the `Workspace` controller from v1 to v2 of the operator.
+
+| Version 1 | Version 2 | Notes |
```
[the remark this answers is not in the document]
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:54]: Thanks!

- Renamed the 3rd column to "Changes between versions". It makes more sense.
- Added `None` to the 3rd column where there are no changes between versions and left `N/A` in the 2nd column when a new version does not have a replacement for this field(removed).

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 16]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| --- | --- | --- |
+| `apiVersion: app.terraform.io/v1alpha1` | `apiVersion: app.terraform.io/v1alpha2` | The `apiVersion` is changed from `v1alpha1` to `v1alpha2`. |
+| `kind: Workspace` | `kind: Workspace` |  |
+| `metadata` | `metadata` |  |
+| `spec.organization` | `spec.organization` |  |
+| `spec.secretsMountPath` | `spec.token.secretKeyRef` | In v2 the operator keeps HCP Terraform access token in a Kubernetes Secret. |
+| `spec.vcs` | `spec.versionControl` | Renamed. Refer to the `versionControl` structure for details. |
+| `spec.vcs.token_id` | `spec.versionControl.oAuthTokenID` | Renamed. Refer to the `versionControl` structure for details. |
```
[the remark this answers is not in the document]
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:62]: Did the same as above: "Renamed the `X` field to `Y`."

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 17]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `spec.vcs.repo_identifier` | `spec.versionControl.repository` | Renamed. Refer to the `versionControl` structure for details. |
+| `spec.vcs.branch` | `spec.versionControl.branch` |  |
```
[the remark this answers is not in the document]
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:64]: Added `None` since the end field didn't change. Thanks!

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 18]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `spec.variables.[*]valueFrom` | `spec.environmentVariables.[*]valueFrom` OR `spec.terraformVariables.[*]valueFrom` |  |
+| `spec.variables.[*]hcl` | `spec.environmentVariables.[*]hcl` OR `spec.terraformVariables.[*]hcl` |  |
+| `spec.variables.sensitive` |  `spec.environmentVariables.[*]sensitive` OR `spec.terraformVariables.[*]sensitive` |  |
+| `spec.variables.environmentVariable` | N/A | Removed, variables are split to `spec.environmentVariables` and `spec.terraformVariables`. |
+| `spec.runTriggers.[*]` | `spec.runTriggers.[*]` |  Changes made to the `runTriggers` structure. |
+| `spec.runTriggers.[*].sourceableName` | `spec.runTriggers.[*].name` | The `sourceableName` field is renamed to `name`. |
+| `spec.sshKeyID` | `spec.sshKey.id` | The `sshKeyID` is moved to `spec.sshKey.id`. |
+| `spec.outputs` | N/A | Removed. |
+| `spec.terraformVersion` | `spec.terraformVersion` |  |
+| `spec.notifications` | `spec.notifications` |  Changes made to the `notification` structure. |
```
[the remark this answers is not in the document]
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:78]: Thank you! Agree, changed it to `None.`

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 19]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `spec.variables.[*]key` | `spec.environmentVariables.[*]name` OR `spec.terraformVariables.[*]name` | The `key` field is renamed to `name`. |
+| `spec.variables.[*]value` | `spec.environmentVariables.[*]value` OR `spec.terraformVariables.[*]value` |  |
+| `spec.variables.[*]valueFrom` | `spec.environmentVariables.[*]valueFrom` OR `spec.terraformVariables.[*]valueFrom` |  |
+| `spec.variables.[*]hcl` | `spec.environmentVariables.[*]hcl` OR `spec.terraformVariables.[*]hcl` |  |
+| `spec.variables.sensitive` |  `spec.environmentVariables.[*]sensitive` OR `spec.terraformVariables.[*]sensitive` |  |
+| `spec.variables.environmentVariable` | N/A | Removed, variables are split to `spec.environmentVariables` and `spec.terraformVariables`. |
+| `spec.runTriggers.[*]` | `spec.runTriggers.[*]` |  Changes made to the `runTriggers` structure. |
```
[the remark this answers is not in the document]
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:73]: I agree with you, changed it to `None.`

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 20]
```diff
@@ -43,33 +43,220 @@ Next, backup the workspace resources.
...
+| `spec.vcs` | `spec.versionControl` | Renamed. Refer to the `versionControl` structure for details. |
+| `spec.vcs.token_id` | `spec.versionControl.oAuthTokenID` | Renamed. Refer to the `versionControl` structure for details. |
+| `spec.vcs.repo_identifier` | `spec.versionControl.repository` | Renamed. Refer to the `versionControl` structure for details. |
+| `spec.vcs.branch` | `spec.versionControl.branch` |  |
+| `spec.vcs.ingress_submodules` | `spec.workingDirectory` | Moved. |
+| `spec.variables.[*]` | `spec.environmentVariables.[*]` OR `spec.terraformVariables.[*]` | Variables were split into two different paths. Depending on the variable's value of `spec.variables.environmentVariable` in v1 CRD:<br/> - If it is `true` then you must migrate the variable to `spec.environmentVariables`.<br/> - If it is `false` then you must migrate the variable to `spec.terraformVariables`. |
+| `spec.variables.[*]key` | `spec.environmentVariables.[*]name` OR `spec.terraformVariables.[*]name` | The `key` field is renamed to `name`. |
+| `spec.variables.[*]value` | `spec.environmentVariables.[*]value` OR `spec.terraformVariables.[*]value` |  |
```
[the remark this answers is not in the document]
Comment (arybolovlev) [reply website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx]: I like this idea. Added an empty `a` tag to the explanation row and `Learn more` pointing to it. I hope I get your idea right. :)

**Comment by arybolovlev** [comment]
> Left style suggestions, but overall is looking great! As always, please feel free to push back on anything that doesn’t feel good to you! ✨

Thank you! I accepted all your suggestions and I like how it looks now. A way better!

I pushed some changes. Could you please take another look?

Thanks!

File: website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx [hunk 21]
```diff
@@ -132,14 +314,28 @@ Specifically, look for an error message such as the following.
...
+If you have any migrated `Module` custom resources, apply them now.
+
+```shell-session
+$ kubectl apply --filename <MIGRATED_V2_MODULE_MANIFEST.yaml>
+```
+
+In v2 of the operator, the `applyMethod` is set to `manual` by default. In this case, a new run in a managed workspace requires manual approval. Run the following command for each `Workspace` resource to change it to `auto` approval.
+
+```shell-session
+kubectl patch workspace <WORKSPACE_NAME> --type=merge --patch '{"spec": {"applyMethod": "auto"}}'
```
Comment (BrianMMcClain) [inline website/docs/cloud-docs/integrations/kubernetes/ops-v2-migration.mdx:340]:
```suggestion
$ kubectl patch workspace <WORKSPACE_NAME> --type=merge --patch '{"spec": {"applyMethod": "auto"}}'
```

**Review by rkoron007 — APPROVED** [review]
Looks good to me! Thank you so much for putting in so much work on this one! ✨
